### ISSUE TITLE
Azure character limits on resource groups

### ISSUE THREAD

**Issue by daxmc99** [rancher/rancher#21061]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
Create a resource Group in azure with a very long  >60 resource name 
Attempt to use that resource group to provision a cluster

**Result:**
Azure cluster bring up returns a error about the name of one of the parameters exceeding a limit specified in https://docs.microsoft.com/en-us/azure/architecture/best-practices/naming-conventions#general 

**Other details that may be helpful:**
Rancher should never send data to azure that we know will exceed a known naming limit
Error seen: 
Operationalinsights.WorkspacesClient#CreateOrUpdate: Invalid input: autorest/validation: validation failed: parameter=workspaceName constraint=MaxLength value="rancher-kube-cluster-dev2-c24b7341-10ad-4e4c-b3f7-f07c483184a9-WEU" details: value length must be less than or equal to 63; waiting on cluster-provisioner-controller (thanks @dbsanfte )

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
2.2.4

**Comment by davidnuzik** [comment]
Duplicate of https://github.com/rancher/rancher/issues/20783 I believe.
We may want to close #20783 in favor of this issue.

**Comment by davidnuzik** [comment]
Version: master-head (6/24/19 9am mst)
I was asked to reproduce the issue.

I see an error like:
`2019/06/24 18:34:42 [ERROR] NodeController c-lnzmj/m-wh9dz [node-controller] failed with : Error creating machine: Error in driver during machine creation: #: : StatusCode=%!!(MISSING)d(<nil>)`
When I attempt to create a cluster with a node template that defines a resource group with a string of longer than 63 chars.
So I am not seeing the exact same error, but perhaps it has something to do with the string used `david-sadgjasdlkfjalsdkjglksadjflkasdjlgkajsdklfjasdlkjgalksdjflkasjdlkgjasldkjfklasdgasdgasdjfkjskdjigjdkfjd`

**Comment by daxmc99** [comment]
Current status:
the issue is caused by this line https://github.com/rancher/rancher/blob/7e115d273fbe56d069fe99aa399aa7e745700569/vendor/github.com/rancher/kontainer-engine/drivers/aks/aks_driver.go#L978
This concatenates 3 values, the resource Group, the subscription id and the workspace region code.
This only occurs when monitoring has been enabled (https://docs.microsoft.com/en-us/azure/aks/view-master-logs).

My current thoughts are that this field should be visible in the UI when monitoring is enabled and contain a unique hint field (likely a hash) if the user opts to use the default value. 

Furthermore, we should clarify the max character limits for user configurable fields. A good rule of thumb should be to keep fields under 63 characters.

**Comment by vincent99** [comment]
The "log analytics workspace" already lives in a particular subscription, so it doesn't seem to me like appending the 36-char ID of it to the name is really accomplishing anything.  All the workspaces in that subscription are going to have the same ID appended.

The region code is possibly helpful since you can presumably have a cluster named "dev" in each AKS region within one subscription.

So I would:

- Set maxLength on Resource Group name to 63 so that the UI limits it correctly if not already the case.

- Take the SubscriptionID (and a hyphen) out of the generated workspaceName entirely (but leave the short region code)

- Check the length of that result.  If it's still >63 chars, switch to a different strategy like `workspaceName=${ substr(workspaceName,0,45) }-${ substr(sha256(workspaceName),0,16) }` so that it is unique but still mostly recognizable to a human.*

*: You could also just truncate it, but that might not be unique and would result in 2 clusters writing to the same workspace since the next block does `CreateOrUpdate(workspaceName)`

**Comment by cjellick** [comment]
pulled back to final review bc we need it backported

**Comment by daxmc99** [comment]
Actual PR for 2.2 branch: https://github.com/rancher/rancher/pull/21235

**Comment by davidnuzik** [comment]
v2.2-head

This is also having the same issue as in master -- there appears to be another limit we are now hitting. When supply Resoruce Group string of 70 chars we get an error like:
`Containerservice.ManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidParameter" Message="The length of the node resource group name is too long. The maximum length is 80 and the length of the value provided is 88. Please see https://aka.ms/aks-naming-rules for more details." Target="name"`

**Comment by davidnuzik** [comment]
**Reproduction:**
Please refer to https://github.com/rancher/rancher/issues/21234#issuecomment-507412994 where this was already reproduced in v2.2.4.

**Validation:**
v2.2-head

- Create AKS cluster with RG string greater than 63 chars, 64 chars in my case. You must have metrics enabled (but this is by default)
`david-abcMHCqGl3L08ME8rk13ISYfc2wD57nORtegG7fks4pX43f3u7FBdmufh1`
- We now successfully create the Log Analytics resource in Azure console, we did not before.
![Screen Shot 2019-07-01 at 3 43 15 PM](https://user-images.githubusercontent.com/45179589/60470892-0b57e000-9c17-11e9-8a7a-056538c67070.png)

- Note: We do get another separate error but this is a new, and separate issue. Azure seems to be appending our RG string with extra characters, this is not something Rancher is doing. Refer to https://github.com/rancher/rancher/issues/21264 for this separate Azure issue.
